### PR TITLE
shane's code

### PR DESCRIPTION
**Pull request by MrCreosote**

### CODE REVIEW

File: scan_all.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
+from pymongo import MongoClient
+import json
+from collections import defaultdict
+import sys
+from checker import checker, GOOD, BAD, UNK, dump_obj
+
+   
+with open('all.json') as f:
```
Comment (MrCreosote) [inline scan_all.py:8]: Is all.json all versions of all objects added since the 26th?

File: scan_all.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    data=f.read()
+objs = json.loads(data)
+
+with open('bad_ids.json') as f:
```
Comment (MrCreosote) [inline scan_all.py:12]: what's in bad_ids?

File: scan_all.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    for h in d['extids']['handle']:
+         h2wsobj[h].append(ref)
+
+with open('ckpt.json') as f:
+    d = f.read()
+good_list = json.loads(d)
+bad_list = []
+all_list = dict()
+good_handles = [ 'KBH_3673745' ] 
+skip = [ '47530/78/1' ]
```
Comment (MrCreosote) [inline scan_all.py:32]: These two lines are cruft I guess?

File: scan_all.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+   
+with open('all.json') as f:
+    data=f.read()
+objs = json.loads(data)
+
+with open('bad_ids.json') as f:
+    data=f.read()
+bad = json.loads(data)
+known = bad.keys()
```
Comment (MrCreosote) [inline scan_all.py:15]: This looks like cruft?

File: scan_all.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
+from pymongo import MongoClient
+import json
+from collections import defaultdict
+import sys
+from checker import checker, GOOD, BAD, UNK, dump_obj
```
Comment (MrCreosote) [inline scan_all.py:5]: Looks like UNK and dump_obj are unused

File: checker.py [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+
+def checker(ref):
+    if not check_missing(ref):
+       return BAD
+    ty = objs[ref]['type']
+    odata = get_shock(ref)
+    if ty.startswith('KBaseGenomes.Genome'):
+        return check_genome(ref, odata)
+    if ty=='Empty.AHandle-1.0':
+        st=UNK
```
Comment (MrCreosote) [inline checker.py:196]: I would say figure out who is saving these and get them to stop. Nuke any that show up as duplicates, and alert the user on any that aren't duplicates.

File: checker.py [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+    elif ty=='KBaseReport.Report-2.0':
+        st=check_report_type(odata)
+    elif ty=='KBaseRNASeq.Bowtie2IndexV2-1.0':
+        st=check_bowtie(odata)   
+    elif ty.startswith('KBaseFile.SingleEndLibrary'):
+        st=check_single(odata)
+    elif ty.startswith('KBaseFile.PairedEndLibrar'):
+        st=check_paired(odata)
+    elif ty.startswith('KBaseAssembly.PairedEndLibrar'):
+        st=check_paired2(odata)
```
Comment (MrCreosote) [inline checker.py:206]: We need to figure out what's saving these and stop it, they should be using KBaseFile

File: checker.py [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+        st=check_assembly(odata, ref)
+    elif ty.startswith('KBaseRNASeq.RNASeqAlignment'):
+        st=check_alignment(odata)
+    elif ty.startswith('KBaseRNASeq.RNASeqExpression'):
+        st=check_expression(odata)
+    elif ty.startswith('KBaseRNASeq.GFFAnnotation'):
+        st=check_gff(odata)
+    elif ty.startswith('KBaseClassifier.GenomeCategorizer'):
+        print("Skipping cat")
+        return GOOD
```
Comment (MrCreosote) [inline checker.py:217]: Why skipping?

File: checker.py [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+    print(fn1)
+    print(fn2)
+    return BAD
+
+def check_cat(odata):
+     print(odata.keys())
+     print(odata['classifier_handle_ref'])
+     print(odata['lib_name'])
+     hid = odata['handle']['hid']
```
Comment (MrCreosote) [inline checker.py:62]: https://narrative.kbase.us/#spec/type/KBaseClassifier.GenomeCategorizer doesn't have a handle field?

File: checker.py [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+def check_report_type(odata):
+    good=0
+    l = odata['file_links']
+    l.extend(odata['html_links'])
+    #l.append(odata['direct_html'])
+    for f in l:
+        if f is None:
+           continue
+        if f=='':
+           continue
```
Comment (MrCreosote) [inline checker.py:87]: Do these 2 cases actually happen? I wouldn't think the type checker would allow it

File: checker.py [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+        if f['handle'] in handles:
```
Comment (MrCreosote) [inline checker.py:88]: not having a handle should be impossible

File: checker.py [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+        kvs = read_kvs(ref)
+        if int(kvs['Number of Protein Encoding Genes']) == gct:
+           return True
+        print(gct)
+        print(odata.keys())
+        print(odata['gff_handle_ref'])
+        print(ref)
+        return BAD
+    fn1=odata['original_source_file_name']
+    hrec = handles[odata['genbank_handle_ref']]
```
Comment (MrCreosote) [inline checker.py:49]: gff_handle_ref is also a handle id

File: checker.py [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+    fn1=odata['original_source_file_name']
+    fn2=hrec['file_name'].replace('.gz','')
+    if fn1==fn2:
```
Comment (MrCreosote) [inline checker.py:52]: I wouldn't trust this check. Some apps could just use the same file name over and over

File: checker.py [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+
+def read_kvs(ref):
+    kvs=dict()
+    for kv in objs[ref]['meta']:
+       kvs[kv['k']] = kv['v']
+    return kvs
+
+def check_genome(ref, odata):
+    if 'original_source_file_name' not in odata:
+        gff=handles[odata['gff_handle_ref']]
```
Comment (MrCreosote) [inline checker.py:38]: genbank_handle_ref is also a handle

File: checker.py [hunk 15]
```diff
@@ -0,0 +1,244 @@
...
+        gct=count_genes(get_shock_file(gff['id']))
+        kvs = read_kvs(ref)
+        if int(kvs['Number of Protein Encoding Genes']) == gct:
+           return True
```
Comment (MrCreosote) [inline checker.py:42]: True -> GOOD?

File: checker.py [hunk 16]
```diff
@@ -0,0 +1,244 @@
...
+def check_missing(ref):
+    for hid in objs[ref]['extids']['handle']:
+        if hid not in handles:
+          return False
+    return True
+
+def get_shock_file(sid):
+    return '/disk1/shock/data/%s/%s/%s/%s/%s.data' % (sid[0:2], sid[2:4], sid[4:6], sid, sid)
+   
+def count_genes(fn):
```
Comment (MrCreosote) [inline checker.py:18]: I'm not familiar enough with GFF format to know if this is correct

File: checker.py [hunk 17]
```diff
@@ -0,0 +1,244 @@
...
+           return GOOD
+       else:
+           return BAD
+
+def check_paired2(odata):
+    good=0
+    print(odata)
+    li = [odata['handle_1']]
+    if 'handle_2' in odata:
+       li.append(odata['handle_2']['file'])
```
Comment (MrCreosote) [inline checker.py:121]: This looks wrong. The first entry of the array will be a `handle_1` and the 2nd will be a `handle_2/file`

File: checker.py [hunk 18]
```diff
@@ -0,0 +1,244 @@
...
+        oid = f['id']
+        id2=handles[hid]['id']
+        if oid==id2:
+            good+=1
+        else:
+            return BAD
+    return GOOD
+
+def check_assembly(odata, ref):
+    if 'fasta_handle_info' not in odata:
```
Comment (MrCreosote) [inline checker.py:148]: There is no `fasta_handle_info` field in assemblies? https://narrative.kbase.us/#spec/type/KBaseGenomeAnnotations.Assembly

The handle fields are:
```
  reads_handle_ref reads_handle_ref;
  fasta_handle_ref fasta_handle_ref;
```

File: checker.py [hunk 19]
```diff
@@ -0,0 +1,244 @@
...
+    good=0
+    oid=odata['file']['id']
+    hid=odata['file']['hid']
+    if hid in handles: 
+        hoid=handles[hid]['id']
+        if hoid==oid:
+            good+=1
+        else:
+            return BAD
+    return GOOD
```
Comment (MrCreosote) [inline checker.py:169]: Why return good if the handle is missing? That seems like the object is broken.

File: scan_all.py (hunk 1 above)
Comment (scanon) [reply scan_all.py:8]: Yes

File: checker.py (hunk 15 above)
Comment (scanon) [reply checker.py:42]: Yes

File: checker.py (hunk 17 above)
Comment (scanon) [reply checker.py:121]: I think this never showed up

**Comment by MrCreosote** [comment]
Moot at this point

**Pull request closed without merging**